lib/pbench/test/unit/agent/task/test_copy_result_tb.py

```diff
@@ -42,9 +34,7 @@ def test_tarball_nonexistent(self, monkeypatch, agent_logger):
         bad_tarball_name = "nonexistent-tarball.tar.xz"
         expected_error_message = f"Tar ball '{bad_tarball_name}' does not exist"
 
-        monkeypatch.setattr(
-            Path, "exists", self.get_path_exists_mock(bad_tarball_name, False)
-        )
+        monkeypatch.setattr(Path, "exists", lambda self: False)
```

Comment:
I freely admit that I'm a big fan of this use of lambda expressions, but I'm curious why you made this change.  The previous code, in addition to providing the `False` value needed to drive the CUT behavior also performed the sanity check that it was being called for the expected file, and this change removes that "blade guard".  What problem was this change intended to solve?

Reply:
```
1@staticmethod
2   def get_path_exists_mock(path: str, result: bool) -> Callable:
3        def mock_func(self: Path) -> bool:
4            assert self.name == path
5            return result
6
7        return mock_func```

It was failing because it was taking file name value as `self.name` value. So, when it was assert in l.4 the `self.name` value that it was taking was `test_copy_result_tb.py` instead of `test_tarball.tar.xz`

Reply:
This is really annoying.

The previous code (which I really think was better...) was working up through build [#5](http://jenkins.intlab.perf-infra.lab.eng.rdu2.redhat.com/job/pbench-ci-multi/job/PR-3220/5/consoleFull) of this PR.

Something changed between Feb 08 05:37 (Jenkins time) and Feb 08 14:36, but it's not easy to tell what.  The `main` branch was updated, but only Dashboard files were changed, but I don't see how that could cause this effect.  I cannot tell what was changed on the PR branch, because the previous commit apparently doesn't exist anymore, but looking over the other changes, I don't see anything there which would cause this, either.

I pulled the branch for this PR, and I can reproduce the problem locally.  However, since it doesn't reproduce on the `main` branch, I would really like to understand what is happening here before we carry this forward.

Reply:
Sure

Reply:
I've purchased a clue.  (Actually, it was handed to me, but it took me awhile to see if for what it was!)

Looking at the failure in build [#6](http://jenkins.intlab.perf-infra.lab.eng.rdu2.redhat.com/job/pbench-ci-multi/job/PR-3220/6/console):
```
...
INTERNALERROR> E               File "/var/tmp/jenkins/tox/agent-py36/lib/python3.6/site-packages/_pytest/python.py", line 1726, in _prunetraceback
INTERNALERROR> E                 path, firstlineno = code.path, code.firstlineno
INTERNALERROR> E               File "/var/tmp/jenkins/tox/agent-py36/lib/python3.6/site-packages/_pytest/_code/code.py", line 95, in path
INTERNALERROR> E                 if not p.exists():
INTERNALERROR> E               File "/var/tmp/jenkins/tox/agent-py36/lib/python3.6/site-packages/pbench/test/unit/agent/task/test_copy_result_tb.py", line 27, in mock_func
INTERNALERROR> E                 assert self.name == path
```

Note the `INTERNALERROR` prefix and the fact that the caller is `_pytest`....  What is happening is that some piece of PyTest itself is calling `Path.exists()` and is hitting our mock!!

So, the assertion that I'm so in love with here is pointing out that there is a problem!  It's not obvious to me how removing that assertion (and always returning, e.g., `False`) solves anything.

We need to figure out how/why our mock is being called unexpectedly and address that.

Reply:
I've narrowed the trigger down to the last assertion in `test_connection_error()`.  The other test scenarios seem to run OK, and that scenario executes past the `with` statement but PyTest loses it during that `assert` statement.

Reply:
OK, that test is failing, and then PyTest is falling over:
```
::TestCopyResults::test_tarball_nonexistent PASSED
::TestCopyResults::test_with_access[public] PASSED
::TestCopyResults::test_with_access[private] PASSED
::TestCopyResults::test_with_access[None] PASSED
::TestCopyResults::test_connection_error 
tce: about to add
tce: about to monkeypatch
tce: about to with
tcb: in with
tce: about to copy
tce: about to assert
tce: excinfo.value="Cannot connect to 'http://pbench.example.com/api/v1/upload/test_tarball.tar.xz': uh-oh"
tce: expected_error_message="Cannot connect to 'http://pbench.example.com/api/v1/upload/test_tarball.tar.xz'"
tce: trying endswidth(): False
```
Should that be `startswith()` and not `endswith()`?

Reply:
> Should that be `startswith()` and not `endswith()`?

The answer is "yes", and, in fact, you'd already made that correction in the most recent commit, but I'd rolled that one back in order to restore the path-exists mocking.

With the changes to `test_copy_result_tb.py`, except for the `startswith()` change, rolled back, the tests now pass.  So, you should be able to restore the path-exists mocking and move forward.

For now, I propose that we keep the more powerful mock and just live with the ungraceful mode of failure, which we will not encounter when things are working.